### PR TITLE
Add default board

### LINKED ISSUE #197: Add default board on first use [$10]
from #12:

- [ ] There should be one board called »Personal« by default from the beginning. Just like in the Calendar app with the first calendar, and for Contacts too with the default address book.
- [ ] There should be default stacks in the first board for sure, and ideally also in every one created: To do, Doing, Done

This might be a good starter issue to get a bit into the php code part. So if anyone wants to give it a try, here is how the implementation should basically look like:

- Implement a class OCA\Deck\Service\DefaultBoardService, methods could be checkFirstRun and createDefaultBoard
- checkFirstRun(): Add a user config value like 'firstrun' to check if the default board has already been created for the current user
- createDefaultBoard(): Create the board, stacks and cards that should appear by default
- check and call createDefaultBoard in PageController::index()

<bountysource-plugin>

---
There is a **[$10 open bounty](https://www.bountysource.com/issues/46408775-add-default-board-on-first-use?utm_campaign=plugin&utm_content=tracker%2F52758431&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F52758431&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Nebri**
This PR should close #197, this is my first PR into an open source project.

If I have missed anything / need revisions please advise upon completing your review. 

Thanks very much, this was a great learning experience!

### CODE REVIEW

File: tests/unit/Service/DefaultBoardServiceTest.php [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+			$this->boardMapper,
+			$this->boardService,
+			$this->stackService,
+			$this->cardService,
+			$this->config
+		);		
+	}
+
+	public function testCheckFirstRunCaseTrue() {
+		$appName = "Deck";
```
Comment (juliushaertl) [inline tests/unit/Service/DefaultBoardServiceTest.php]: Always use single quotes please :wink:

File: tests/unit/Service/DefaultBoardServiceTest.php [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+	public function setUp() {
+		parent::setUp();
+		$this->boardMapper = $this->createMock(BoardMapper::class);
+		$this->boardService = $this->createMock(BoardService::class);
+		$this->stackService = $this->createMock(StackService::class);
+		$this->cardService = $this->createMock(CardService::class);
+		$this->config = $this->createMock(IConfig::class);
+		
+		$this->l10n = $this->request = $this->getMockBuilder(
```
Comment (juliushaertl) [inline tests/unit/Service/DefaultBoardServiceTest.php]: You can also use createMock here.

File: tests/unit/Service/DefaultBoardServiceTest.php [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+		$this->config->expects($this->once())
+			->method('setUserValue');			
+
+		$result = $this->service->checkFirstRun($this->userId, $appName);
+		$this->assertEquals($result, true);
+	}
+
+	public function testCheckFirstRunCaseFalse() {
+		$appName = "deck";
```
Comment (juliushaertl) [inline tests/unit/Service/DefaultBoardServiceTest.php]: Single quotes

File: tests/unit/controller/PageControllerTest.php [hunk 4]
```diff
@@ -42,16 +48,43 @@ public function setUp() {
...
 		$this->controller = new PageController(
-			'deck', $this->request, $this->l10n, $this->userId
+			'deck', $this->request, $this->defaultBoardService, $this->l10n, $this->userId
 		);
 	}
+	
+	public function testIndexOnFirstRun() {
+
+		$board = new Board();
+		$board->setTitle("Personal");
```
Comment (juliushaertl) [inline tests/unit/controller/PageControllerTest.php]: single quotes

File: lib/Service/DefaultBoardService.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		$this->boardService = $boardService;
+		$this->stackService = $stackService;
+		$this->cardService = $cardService;
+		$this->config = $config;
+		$this->boardMapper = $boardMapper;
+		$this->l10n = $l10n;
+    }
+    
+    public function checkFirstRun($userId, $appName) {        
+		$firstRun = $this->config->getUserValue($userId,$appName,'firstRun','yes');
```
Comment (juliushaertl) [inline lib/Service/DefaultBoardService.php]: Please add a space after the ,

File: lib/Service/DefaultBoardService.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		$userBoards = $this->boardMapper->findAllByUser($userId);
+		
+		if ($firstRun === 'yes' && count($userBoards) === 0) {
+			$this->config->setUserValue($userId,$appName,'firstRun','no');
```
Comment (juliushaertl) [inline lib/Service/DefaultBoardService.php]: space after ,

File: tests/unit/controller/PageControllerTest.php [hunk 7]
```diff
@@ -42,16 +48,43 @@ public function setUp() {
...
 
-	public function testIndex() {
 		$response = $this->controller->index();
-		$this->assertEquals('main', $response->getTemplateName());
+		$this->assertEquals('main', $response->getTemplateName());		
 	}
 
+	public function testIndexOnSecondRun() {
+
+		$this->config->setUserValue($this->userId,'deck','firstRun','no');
```
Comment (juliushaertl) [inline tests/unit/controller/PageControllerTest.php]: space after ,

File: l10n/en_GB.js [hunk 8]
```diff
@@ -74,6 +74,6 @@ OC.L10N.register(
     "Shared boards" : "Shared boards",
     "View more" : "View more",
     "Move board to archive" : "Move board to archive",
-    "Create a new board" : "Create a new board"
+	"Create a new board" : "Create a new board"
```
Comment (juliushaertl) [inline l10n/en_GB.js]: This change can be removed as well.

File: l10n/en_GB.json [hunk 9]
```diff
@@ -72,6 +72,6 @@
     "Shared boards" : "Shared boards",
     "View more" : "View more",
     "Move board to archive" : "Move board to archive",
-    "Create a new board" : "Create a new board"
+	"Create a new board" : "Create a new board"
```
Comment (juliushaertl) [inline l10n/en_GB.json]: Same here.

**Comment by juliushaertl** [comment]
:+1: for adding tests. There seems to be one still failing: 

```

There was 1 failure:

1) OCA\Deck\Service\DefaultBoardServiceTest::testCreateDefaultBoard
Expectation failed for method name is equal to <string:create> when invoked 3 time(s)
Parameter 0 for invocation #0 OCA\Deck\Service\StackService::create(null, 5, 1) does not match expected value.
Failed asserting that null matches expected 'To do'.

/drone/src/github.com/nextcloud/server/apps/deck/lib/Service/DefaultBoardService.php:78
/drone/src/github.com/nextcloud/server/apps/deck/tests/unit/Service/DefaultBoardServiceTest.php:168
```

**Comment by Nebri** [comment]
> There was 1 failure:

>1) OCA\Deck\Service\DefaultBoardServiceTest::testCreateDefaultBoard
Expectation failed for method name is equal to <string:create> when invoked 3 time(s)

>Parameter 0 for invocation #0 OCA\Deck\Service\StackService::create(null, 5, 1) does not match expected value.
>Failed asserting that null matches expected 'To do'.

>/drone/src/github.com/nextcloud/server/apps/deck/lib/Service/DefaultBoardService.php:78
>/drone/src/github.com/nextcloud/server/apps/deck/tests/unit/Service/DefaultBoardServiceTest.php:168

Part of this test is checking that the name of the stacks is being set, and those names are going through l10n->t which does not yet have a definition for those strings, hence why the test fails.

Should I be removing that part of the test?

File: l10n/en_GB.js (hunk 8 above)
Comment (Nebri) [reply l10n/en_GB.js]: Isn't this string part of the l10n definitions? I see it in every other file, all I did here was remove the comma I appended at the end of the line, and obvious I accidentally added in some extra white space.

Comment (juliushaertl) [reply l10n/en_GB.js]: I was referring to the whitespace :wink: But that's a non-blocker.

File: tests/unit/Service/DefaultBoardServiceTest.php [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+		$board->setId($boardId);
+		$board->setTitle($title);
+		$board->setOwner($this->userId);
+		$board->setColor($color);
+		$this->boardService->expects($this->once())
+			->method('create')
+			->willReturn($board);
+
+		$stackToDoId = '123';		
+		$stackToDo = $this->assembleTestStack($this->l10n->t('To do'), $stackToDoId, $boardId);
```
Comment (juliushaertl) [inline tests/unit/Service/DefaultBoardServiceTest.php]: If you mock the t call on l10n with something like:
```
$this->l10n->expects($this->any())
	->method('t')
	->willReturnCallback(function($text) { return $text } );
```

you can just call $this->assembleTestStack($this->l10n->t('

Comment (Nebri) [reply tests/unit/Service/DefaultBoardServiceTest.php]: Great, just submitted the revision, once your build server finishes it's thing you should be good to go :).

**Comment by juliushaertl** [comment]
Sign off check failing due to github branch update. Let's merge this.

**Comment by juliushaertl** [comment]
@Nebri Congratulations to your first open source contribution. :tada: I'm glad to have you on board for the deck app.

**Pull request merged**